File: services/payment-orchestrator-java/src/main/java/com/acme/payments/orchestrator/infrastructure/ApiKeyAuthenticationFilter.java
package com.acme.payments.orchestrator.infrastructure;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;

@Component
public class ApiKeyAuthenticationFilter extends OncePerRequestFilter {
    private static final String API_KEY_HEADER = "X-Internal-Api-Key";

    private final String apiKey;

    public ApiKeyAuthenticationFilter(@Value("${payments.security.api-key:}") String apiKey) {
        this.apiKey = apiKey;
    }

    @Override
    protected boolean shouldNotFilter(HttpServletRequest request) {
        return apiKey == null || apiKey.isBlank() || request.getRequestURI().startsWith("/actuator");
    }

    @Override
    protected void doFilterInternal(
            HttpServletRequest request,
            HttpServletResponse response,
            FilterChain filterChain
    ) throws ServletException, IOException {
        if (!apiKey.equals(request.getHeader(API_KEY_HEADER))) {
            response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "invalid internal api key");
            return;
        }

        filterChain.doFilter(request, response);
    }
}

File: services/pix-boleto-kotlin/src/main/kotlin/com/acme/payments/pixboleto/infrastructure/ApiKeyAuthenticationFilter.kt
package com.acme.payments.pixboleto.infrastructure

import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

@Component
class ApiKeyAuthenticationFilter(
    @Value("\${payments.security.api-key:}") private val apiKey: String,
) : OncePerRequestFilter() {
    override fun shouldNotFilter(request: HttpServletRequest): Boolean {
        return apiKey.isBlank() || request.requestURI.startsWith("/actuator")
    }

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain,
    ) {
        if (request.getHeader(API_KEY_HEADER) != apiKey) {
            response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "invalid internal api key")
            return
        }

        filterChain.doFilter(request, response)
    }

    private companion object {
        const val API_KEY_HEADER = "X-Internal-Api-Key"
    }
}
